sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -0,0 +1,230 @@
+        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
+        /// <param name="partitionProcessorFactory">A factory used to create partition processors.  Its implementation must be provided by the user.</param>
+        /// <param name="partitionManager">A partition manager used to interact with the storage system.  Its implementation must be provided by the user.</param>
+        /// <param name="options">The set of options to use for this event processor.</param>
+        ///
+        public EventProcessor(string consumerGroup,
+                              EventHubClient eventHubClient,
+                              IPartitionProcessorFactory partitionProcessorFactory,
+                              IPartitionManager partitionManager,
+                              EventProcessorOptions options)
+        {
+            Guard.ArgumentNotNull(nameof(eventHubClient), eventHubClient);
+            Guard.ArgumentNotNullOrEmpty(nameof(consumerGroup), consumerGroup);
+            Guard.ArgumentNotNull(nameof(partitionProcessorFactory), partitionProcessorFactory);
+            Guard.ArgumentNotNull(nameof(partitionManager), partitionManager);
+
+            InnerClient = eventHubClient;
+            ConsumerGroup = consumerGroup;
+            PartitionProcessorFactory = partitionProcessorFactory;
+            PartitionManager = partitionManager;
+            Options = options?.Clone() ?? new EventProcessorOptions();
+
+            InstanceId = Guid.NewGuid().ToString();
+        }
+
+        /// <summary>
+        ///   Starts the event processor.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Start()
+        {
+            if (RunningTask == null)
+            {
+                RunningTaskTokenSource = new CancellationTokenSource();
+
+                PartitionPumps = new ConcurrentDictionary<string, PartitionPump>();
+
+                var partitionIds = await InnerClient.GetPartitionIdsAsync();
+
+                await Task.WhenAll(partitionIds.Select(async partitionId =>
+                {
+                    var partitionContext = new PartitionContext(InnerClient.EventHubPath, ConsumerGroup, InstanceId, partitionId);
+                    var checkpointManager = new CheckpointManager(partitionContext, PartitionManager);
+
+                    var partitionProcessor = PartitionProcessorFactory.CreatePartitionProcessor(partitionContext, checkpointManager);
+
+                    var partitionPump = new PartitionPump(InnerClient, ConsumerGroup, partitionId, partitionProcessor, Options);
+                    PartitionPumps.TryAdd(partitionId, partitionPump);
+
+                    await partitionPump.Start();
+                }));
+
+                RunningTask = Run(RunningTaskTokenSource.Token);
```

Comment:
Are you sure this is actually executing your task?   I may be misreading, but if you're not awaiting it or explicitly starting it via `Task.Factory.StartNew` or `Task.Run`, then I don't believe that you're actually executing.   Do you have a test to prove this works the way that you're intending?

Reply:
I haven't tried reaching this path of code (`PartitionPump` failure), but `RunningTask = Run(something)` works when starting the `PartitionPump`.

Reply:
Just reached this part of the code (due to some unexpected errors when addressing the comments) and I can confirm it works as expected.